src/uv_fs.c

```diff
@@ -264,10 +267,87 @@ int UvFsMakeFile(const char *dir,
                  unsigned n_bufs,
                  char *errmsg)
 {
+    int rv;
+    char tmp_filename[UV__FILENAME_LEN+1] = {0};
+    char path[UV__PATH_SZ] = {0};
+    char tmp_path[UV__PATH_SZ] = {0};
+
+    /* Create a temp file with the given content
+     * TODO as of libuv 1.34.0, use `uv_fs_mkstemp` */
+    size_t sz = sizeof(tmp_filename);
+    rv = snprintf(tmp_filename, sz, TMP_FILE_FMT, filename);
+    if (rv < 0 || rv >= (int) sz) {
+        return rv;
+    }
     int flags = UV_FS_O_WRONLY | UV_FS_O_CREAT | UV_FS_O_EXCL;
-    return uvFsWriteFile(dir, filename, flags, bufs, n_bufs, errmsg);
+    rv = uvFsWriteFile(dir, tmp_filename, flags, bufs, n_bufs, errmsg);
+    if (rv != 0) {
+        goto err_after_tmp_create;
+    }
+
+    /* Check if the file exists */
+    bool exists = false;
+    rv = UvFsFileExists(dir, filename, &exists, errmsg);
+    if (rv != 0) {
+        goto err_after_tmp_create;
+    }
+    if (exists) {
+        rv = -1;
+        goto err_after_tmp_create;
+    }
+
+    /* Rename the temp file. Remark that there is a race between the existence
+     * check and the rename, there is no `renameat2` equivalent in libuv.
+     * However, in the current implementation this should pose no problems.*/
+    UvOsJoin(dir, tmp_filename, tmp_path);
+    UvOsJoin(dir, filename, path);
+    rv = UvOsRename(tmp_path, path);
```

Comment:
Perhaps it's worth calling `UvFsSyncDir` too, to ensure the directory metadata is written on disk. YMMV though